# Extraccion de los datos 
 Se descomprime el json comprimido usando la funcion nativa de Pandas 2.0 read_json con los parametros. Luego se almacena en una variable de tipo dataframe con los datos brutos provenientes del json descomprimido

In [46]:
import numpy as np
import pandas as pd
import gzip
import os
import json
import ast

# Ruta al dataset de Juegos Steam
path = '../dataset/steam_games.json.gz'

data = []
with gzip.open(path, 'r') as f:
    for line in f:
      data.append(json.loads(line))

# Convertimos a DataFrame
df_games = pd.DataFrame(data)
df_games



publisher                                 genres  \
0                   NaN                                    NaN   
1                   NaN                                    NaN   
2                   NaN                                    NaN   
3                   NaN                                    NaN   
4                   NaN                                    NaN   
...                 ...                                    ...   
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444              NaN                                    NaN   

                        app_name                     title  \
0                            NaN                       NaN   
1                            NaN                       NaN   
2                            NaN                       NaN   
3                            NaN                       NaN   
4                            NaN                       NaN   
...                          ...                       ...   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                       NaN   

                                                      url release_date  \
0                                                     NaN          NaN   
1                                                     NaN          NaN   
2                                                     NaN          NaN   
3                                                     NaN          NaN   
4                                                     NaN          NaN   
...                                                   ...          ...   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                     tags  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440  http://steamcommunity.com/app/773640/reviews/?...   
120441  http://steamcommunity.com/app/733530/reviews/?...   
120442  http://steamcommunity.com/app/610660/reviews/?...   
120443  http://steamcommunity.com/app/658870/reviews/?...   
120444  http://steamcommunity.com/app/681550/reviews/?...   

                                                    specs price early_access  \
0                                        

Hacemos un resumen consiso de los Dataframe

In [47]:
df_games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120445 entries, 0 to 120444
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher     24083 non-null  object
 1   genres        28852 non-null  object
 2   app_name      32133 non-null  object
 3   title         30085 non-null  object
 4   url           32135 non-null  object
 5   release_date  30068 non-null  object
 6   tags          31972 non-null  object
 7   reviews_url   32133 non-null  object
 8   specs         31465 non-null  object
 9   price         30758 non-null  object
 10  early_access  32135 non-null  object
 11  id            32133 non-null  object
 12  developer     28836 non-null  object
dtypes: object(13)
memory usage: 11.9+ MB


In [48]:
round(df_games.isna().sum().to_frame('% de Nulos')*100/len(df_games),2)

% de Nulos
publisher          80.00
genres             76.05
app_name           73.32
title              75.02
url                73.32
release_date       75.04
tags               73.46
reviews_url        73.32
specs              73.88
price              74.46
early_access       73.32
id                 73.32
developer          76.06

In [49]:
# Describe todas las columnas del dataframe df_games
df_games.describe()


publisher    genres    app_name       title  \
count      24083     28852       32133       30085   
unique      8239       883       32094       30054   
top      Ubisoft  [Action]  Soundtrack  Soundtrack   
freq         385      1880           3           3   

                                                      url release_date  \
count                                               32135        30068   
unique                                              32135         3582   
top     http://store.steampowered.com/app/761140/Lost_...   2012-10-16   
freq                                                    1          100   

                        tags  \
count                  31972   
unique                 15395   
top     [Casual, Simulation]   
freq                    1292   

                                              reviews_url            specs  \
count                                               32133            31465   
unique                                              32132             4649   
top     http://steamcommunity.com/app/612880/reviews/?...  [Single-player]   
freq                                                    2             2794   

           price early_access      id                developer  
count   30758.00        32135   32133                    28836  
unique    162.00            2   32132                    10992  
top         4.99        False  612880  Ubisoft - San Francisco  
freq     4278.00        30188       2                     1259

#### Descripción de los datos.
A partir del análisis exploratorio anterior, podemos observar que el conjunto de datos contiene 120445 registros y 13 variables con información sobre juegos de Steam. Sin embargo, hay una gran cantidad de datos nulos (75% en promedio). A continuación, se describen las variables:


- publisher: empresa publicadora del juego
- genres: género del juego. Esta formado por una lista de uno o mas géneros por registro
- app_name: nombre del juego
- title: título del juego
- url: URL de publicación del juego
- release_date: fecha de lanzamiento en formato 2018-01-04
- tags: etiquetas del juego. Esta formado por una lista de uno o más etiquetas por registro
- reviews_url: reviews del juego
- specs: especificaciones. Es una lista con uno o mas string con las especificaciones
- price: precio del juego
- early_access: indica el acceso temprano al juego con un True/False
- id: identificador único del juego
- developer: desarrollador del juego

# Transformación
En esta sección, realizamos la limpieza inicial de los datos y las transformaciones necesarias. Esto puede incluir la creación de nuevas columnas a partir de las existentes, la eliminación de duplicados o columnas innecesarias, la gestión de valores nulos o la corrección de tipos de datos.

In [50]:
df_games.dropna(subset='id', inplace=True)
df_games.reset_index(drop=True, inplace=True)
df_games.shape

(32133, 13)

#### Verificación de duplicados
Del análisis exploratorio anterior notamos que el id 612880 esta duplicado. Se procede a verificar:

In [51]:
df_games[df_games['id'].duplicated(keep=False)]

publisher    genres                          app_name  \
13893  Bethesda Softworks  [Action]  Wolfenstein II: The New Colossus   
14572  Bethesda Softworks  [Action]  Wolfenstein II: The New Colossus   

                                  title  \
13893  Wolfenstein II: The New Colossus   
14572  Wolfenstein II: The New Colossus   

                                                     url release_date  \
13893          http://store.steampowered.com/app/612880/   2017-10-26   
14572  http://store.steampowered.com/app/612880/Wolfe...   2017-10-26   

                                                    tags  \
13893  [Action, FPS, Gore, Violent, Alternate History...   
14572  [Action, FPS, Gore, Violent, Alternate History...   

                                             reviews_url  \
13893  http://steamcommunity.com/app/612880/reviews/?...   
14572  http://steamcommunity.com/app/612880/reviews/?...   

                                                   specs  price early_access  \
13893  [Single-player, Steam Achievements, Full contr...  59.99        False   
14572  [Single-player, Steam Achievements, Full contr...  59.99        False   

           id      developer  
13893  612880  Machine Games  
14572  612880  Machine Games

In [52]:
# Eliminamos duplicados por id
df_games.drop_duplicates(subset='id', inplace=True)
df_games.shape

(32132, 13)

In [53]:
# Renombra la columna 'id' a 'item_id' en el DataFrame df_games
df_games.rename(columns={'id': 'item_id'}, inplace=True)
df_games.columns

Index(['publisher', 'genres', 'app_name', 'title', 'url', 'release_date',
       'tags', 'reviews_url', 'specs', 'price', 'early_access', 'item_id',
       'developer'],
      dtype='object')

#### Columnas title y app_name
##### Gestión de nulos
Notamos que ambas columnas contienen valores nulos que procedemos a reemplazar con 'unknown'.

In [54]:
columns = ['title', 'app_name']
df_games[columns] = df_games[columns].fillna('unknown')
df_games.isnull().sum()

publisher       8051
genres          3282
app_name           0
title              0
url                0
release_date    2066
tags             162
reviews_url        0
specs            669
price           1377
early_access       0
item_id            0
developer       3298
dtype: int64

In [55]:
# Realizamos la comparación y creamos una columna con el resultado.
df_games['is_equal'] = df_games['app_name'] == df_games['title']

# Calculamos el porcentaje de filas donde 'app_name' y 'title' son iguales
equal_percentage = round((df_games['is_equal'].sum() / len(df_games)) * 100, 2)

print(f"El porcentaje de filas donde 'app_name' y 'title' son iguales es: {equal_percentage}%")

El porcentaje de filas donde 'app_name' y 'title' son iguales es: 91.9%


In [56]:
# Filtramos las filas donde 'app_name' y 'title' no son iguales.
not_equal_df = df_games.loc[df_games['is_equal'] == False]
not_equal_df[['app_name', 'title']]

app_name    title
4             Log Challenge  unknown
11     Icarus Six Sixty Six  unknown
19            After Life VR  unknown
20               Kitty Hawk  unknown
22               Mortars VR  unknown
...                     ...      ...
32071        Tank of War-VR  unknown
32074           Flappy Arms  unknown
32075           SpaceWalker  unknown
32083            LIV Client  unknown
32132           Maze Run VR  unknown

[2603 rows x 2 columns]

Cambiamos el nombre de la columna <code>app_name</code> por uno mas representativo como <code>item_name</code>.

In [57]:
df_games.rename(columns={'app_name': 'item_name'}, inplace=True)
df_games.columns

Index(['publisher', 'genres', 'item_name', 'title', 'url', 'release_date',
       'tags', 'reviews_url', 'specs', 'price', 'early_access', 'item_id',
       'developer', 'is_equal'],
      dtype='object')

##### Tratamiento Columas <code>publisher</code>
Gestión de nulos
Notamos que también tenemos valores nulos para la columnas publisher y developer que reemplazamos con 'unknown'.

In [58]:
# Reemplazamos los valores nulos con 'unknown'
columns = ['publisher', 'developer']
df_games[columns] = df_games[columns].fillna('unknown')
df_games.isnull().sum()

publisher          0
genres          3282
item_name          0
title              0
url                0
release_date    2066
tags             162
reviews_url        0
specs            669
price           1377
early_access       0
item_id            0
developer          0
is_equal           0
dtype: int64

##### Tratamiento de <code>release_date</code>
Valor de nulos

In [59]:
df_games['release_date'] = df_games['release_date'].fillna('unknown')
df_games.isnull().sum()

publisher          0
genres          3282
item_name          0
title              0
url                0
release_date       0
tags             162
reviews_url        0
specs            669
price           1377
early_access       0
item_id            0
developer          0
is_equal           0
dtype: int64

In [60]:
import re
mask = df_games['release_date'].apply(lambda x: not re.match(r'\d{4}-\d{2}-\d{2}', str(x)))

- Filtramos las filas donde la máscara es True.

In [61]:
invalid_dates = df_games.loc[mask, 'release_date']
invalid_dates

4             unknown
10             Soon..
11            unknown
19            unknown
20            unknown
             ...     
32085     Coming Soon
32086         unknown
32095            2016
32121    January 2018
32132         unknown
Name: release_date, Length: 2351, dtype: object

<h1>Extracción del año.</h1>
Para los endpoints que se consumirán en la API solo necesitamos el año de lanzamiento del item, para ello lo extraemos de release_date de la siguiente manera:

- Creamos una máscara booleana donde release_date no coincide con el formato YYYY-MM-DD.

In [62]:
mask = df_games['release_date'].apply(lambda x: not re.match(r'\d{4}-\d{2}-\d{2}', str(x)))

# Filtramos las filas donde la máscara es True
invalid_dates = df_games.loc[mask, 'release_date']
invalid_dates

4             unknown
10             Soon..
11            unknown
19            unknown
20            unknown
             ...     
32085     Coming Soon
32086         unknown
32095            2016
32121    January 2018
32132         unknown
Name: release_date, Length: 2351, dtype: object

In [63]:
# Extraemos el año en una nueva columna y si no existe reemplazamos por 'unknown'.
df_games['release_year'] = df_games['release_date'].str.extract(r'(\d{4})').fillna('unknown')
df_games['release_year']

0           2018
1           2018
2           2017
3           2017
4        unknown
          ...   
32128       2018
32129       2018
32130       2018
32131       2017
32132    unknown
Name: release_year, Length: 32132, dtype: object

In [64]:
# Comprobamos que se haya extraido correctamente el año:
df_games[['release_date', 'release_year']].loc[[32121, 32095, 32085 ]]

release_date release_year
32121  January 2018         2018
32095          2016         2016
32085   Coming Soon      unknown

##### Buscamos años inexistentes en <code>release_year</code> y los pasamos a 'unknown'.

In [65]:
df_games['release_year'].unique()

array(['2018', '2017', 'unknown', '1997', '1998', '2016', '2006', '2005',
       '2003', '2007', '2002', '2000', '1995', '1996', '1994', '2001',
       '1993', '2004', '1999', '2008', '2009', '1992', '1989', '2010',
       '2011', '2013', '2012', '2014', '1983', '1984', '2015', '1990',
       '1988', '1991', '1985', '1982', '1987', '1981', '1986', '2021',
       '5275', '2019', '1975', '1970', '1980'], dtype=object)

In [66]:
df_games[df_games['release_year'] == "5275"]

publisher                         genres            item_name  \
13427    一次元创作组  [Casual, Indie, Early Access]  Puzzle Sisters Foer   

                     title                                                url  \
13427  Puzzle Sisters Foer  http://store.steampowered.com/app/710190/Puzzl...   

      release_date                           tags  \
13427  预热群52756441  [Early Access, Casual, Indie]   

                                             reviews_url  \
13427  http://steamcommunity.com/app/710190/reviews/?...   

                                                   specs price early_access  \
13427  [Single-player, Steam Achievements, Steam Trad...   NaN         True   

      item_id developer  is_equal release_year  
13427  710190    一次元创作组      True         5275

In [67]:
df_games['release_year'] = df_games['release_year'].replace('5275', 'unknown')
df_games['release_year'].unique()

array(['2018', '2017', 'unknown', '1997', '1998', '2016', '2006', '2005',
       '2003', '2007', '2002', '2000', '1995', '1996', '1994', '2001',
       '1993', '2004', '1999', '2008', '2009', '1992', '1989', '2010',
       '2011', '2013', '2012', '2014', '1983', '1984', '2015', '1990',
       '1988', '1991', '1985', '1982', '1987', '1981', '1986', '2021',
       '2019', '1975', '1970', '1980'], dtype=object)

#### Columnas de Precio
- Gestión de nulos

In [68]:
df_games[df_games['price'].isnull()][['price','genres', 'tags']].sample(10)

price                                 genres  \
26671   NaN                                  [RPG]   
14771   NaN                [Action, Casual, Indie]   
1824    NaN                    [Action, Adventure]   
31213   NaN                           [Simulation]   
26295   NaN                          [Action, RPG]   
3754    NaN                 [Casual, Free to Play]   
6653    NaN                        [Casual, Indie]   
25377   NaN         [Action, Simulation, Strategy]   
12394   NaN                                    NaN   
26790   NaN  [Action, Indie, Simulation, Strategy]   

                                                    tags  
26671                                              [RPG]  
14771                            [Action, Indie, Casual]  
1824                                 [Action, Adventure]  
31213                                                NaN  
26295                 [Action, RPG, Co-op, Stealth, FPS]  
3754                              [Casual, Free to Play]  
6653                                     [Indie, Casual]  
25377  [Strategy, Simulation, Action, World War I, Hi...  
12394  [Design & Illustration, Utilities, Animation &...  
26790              [Strategy, Action, Indie, Simulation]

In [69]:
df_games.price.isnull().sum()

1377

In [70]:
# Crea una función para comprobar si un juego es gratis
def check_free(val):
    if isinstance(val, list):
        for item in val:
            if 'free' in item.lower():
                return True
    return False


# Aplicamos la función a cada fila del DataFrame
df_games['is_free'] = df_games['genres'].apply(check_free) | df_games['tags'].apply(check_free) | df_games['specs'].apply(check_free)

# Si un juego es gratuito y su precio es 'NaN', establece su precio en 0
df_games.loc[(df_games['is_free'] == True) & (df_games['price'].isnull()), 'price'] = 0

# Eliminamos la columna temporal 'is_free'
df_games.drop('is_free', axis=1, inplace=True)

Observamos que la cantidad de valores nulos ha disminuido, por lo que podemos concluir que teníamos varios juegos gratuitos con precios nulos, pero aún nos falta asignarle precio 0 a los juegos cuyo precio indica que es 'free' en la columna price.

In [71]:
df_games.price.isnull().sum()

1172

In [72]:
def replace_free_with_zero(val):
    if isinstance(val, str) and 'free' in val.lower():
        return 0
    return val

# Solo aplica la función a la columna 'price'
df_games['price'] = df_games['price'].apply(replace_free_with_zero)

In [73]:
df_games['price'].unique()

array([4.99, 0, 0.99, 2.99, 3.99, 9.99, 18.99, 29.99, nan, 10.99, 1.59,
       14.99, 1.99, 59.99, 8.99, 6.99, 7.99, 39.99, 19.99, 7.49, 12.99,
       5.99, 2.49, 15.99, 1.25, 24.99, 17.99, 61.99, 3.49, 11.99, 13.99,
       34.99, 74.76, 1.49, 32.99, 99.99, 14.95, 69.99, 16.99, 79.99,
       49.99, 5.0, 44.99, 13.98, 29.96, 119.99, 109.99, 149.99, 771.71,
       'Install Now', 21.99, 89.99, 'Play WARMACHINE: Tactics Demo', 0.98,
       139.92, 4.29, 64.99, 54.99, 74.99, 'Install Theme', 0.89,
       'Third-party', 0.5, 'Play Now', 299.99, 1.29, 3.0, 15.0, 5.49,
       23.99, 49.0, 20.99, 10.93, 1.39, 36.99, 4.49, 2.0, 4.0, 9.0,
       234.99, 1.95, 1.5, 199.0, 189.0, 6.66, 27.99, 10.49, 129.99, 179.0,
       26.99, 399.99, 31.99, 399.0, 20.0, 40.0, 3.33, 199.99, 22.99,
       320.0, 38.85, 71.7, 59.95, 995.0, 27.49, 3.39, 6.0, 19.95, 499.99,
       16.06, 4.68, 131.4, 44.98, 202.76, 1.0, 2.3, 0.95, 172.24, 249.99,
       2.97, 10.96, 10.0, 30.0, 2.66, 6.48, 19.29, 11.15, 18.9, 2.89,
  

In [74]:
non_price_values = ['Install Now', 'Play WARMACHINE: Tactics Demo', 'Install Theme', 'Third-party', 'Play Now', 'Play the Demo', 'Starting at $499.00', 'Starting at $449.00']
df_games[df_games['price'].isin(non_price_values)]

publisher  \
2404                          EVGA   
2870   Privateer Press Interactive   
3831                       unknown   
3917                       unknown   
4025                     DigitalEZ   
22733        Boomzap Entertainment   
24999                      unknown   
25000                      unknown   
26216                      unknown   
31836           PopCap Games, Inc.   

                                            genres  \
2404                                   [Utilities]   
2870                             [Indie, Strategy]   
3831   [Adventure, Casual, Indie, RPG, Simulation]   
3917                                       [Indie]   
4025                               [Casual, Indie]   
22733                                   [Strategy]   
24999                                          NaN   
25000                                          NaN   
26216       [Adventure, Casual, Indie, Simulation]   
31836                                     [Casual]   

                         item_name                        title  \
2404            EVGA Precision XOC           EVGA Precision XOC   
2870           WARMACHINE: Tactics          WARMACHINE: Tactics   
3831         FREE China Theme Pack        FREE China Theme Pack   
3917           Parcel - Soundtrack          Parcel - Soundtrack   
4025   Oblivious Garden ~White Day  Oblivious Garden ~White Day   
22733          Legends of Callasia          Legends of Callasia   
24999          Syber Steam Machine          Syber Steam Machine   
25000      Alienware Steam Machine      Alienware Steam Machine   
26216       Area-X - Extra Gallery       Area-X - Extra Gallery   
31836               Peggle Extreme               Peggle Extreme   

                                                     url release_date  \
2404   http://store.steampowered.com/app/268850/EVGA_...   2014-09-19   
2870   http://store.steampowered.com/app/253510/WARMA...   2014-11-20   
3831   http://store.steampowered.com/app/370880/FREE_...   2015-06-10   
3917   http://store.steampowered.com/app/362970/Parce...   2015-07-02   
4025   http://store.steampowered.com/app/345040/Obliv...   2015-07-20   
22733  http://store.steampowered.com/app/438920/Legen...   2016-06-10   
24999  http://store.steampowered.com/app/353420/Syber...   2015-11-10   
25000  http://store.steampowered.com/app/353390/Alien...   2015-11-10   
26216  http://store.steampowered.com/app/383860/AreaX...   2015-06-24   
31836  http://store.steampowered.com/app/3483/Peggle_...   2007-09-11   

                                                    tags  \
2404                 [Utilities, Software, Free to Play]   
2870   [Strategy, Turn-Based, Turn-Based Strategy, St...   
3831         [Adventure, RPG, Indie, Casual, Simulation]   
3917                                             [Indie]   
4025                                     [Casual, Indie]   
22733  [Strategy, Wargame, Fantasy, Multiplayer, Turn...   
24999                          [Steam Machine, Hardware]   
25000  [Steam Machine, Hardware, Gaming, Futuristic, ...   
26216             [Adventure, Indie, Casual, Simulation]   
31836             [Casual, Puzzle, Free to Play, Action]   

                                             reviews_url  \
2404   http://steamcommunity.com/app/268850/reviews/?...   
2870   http://steamcommunity.com/app/253510/reviews/?...   
3831   http://steamcommunity.com/app/370880/reviews/?...   
3917   http://steamcommunity.com/app/362970/reviews/?...   
4025   http://steamcommunity.com/app/345040/reviews/?...   
22733  http://steamcommunity.com/app/438920/reviews/?...   
24999  http://steamcommunity.com/app/353420/reviews/?...   
25000  http://steamcommunity.com/app/353390/reviews/?...   
26216  http://steamcommunity.com/app/383860/reviews/?...   
31836  http://steamcommunity.com/app/3483/reviews/?br...   

                                                   specs  \
2404                 [Single-player, Steam Achievements]   
2870   [Single-player, Mult

Transformamos los valores que no representan un precio en 0.

In [75]:
df_games.loc[df_games['price'] == 'Starting at $499.00', 'price'] = 499
df_games.loc[df_games['price'] == 'Starting at $449.00', 'price'] = 449
non_price_values = ['Install Now', 'Play WARMACHINE: Tactics Demo', 'Install Theme', 'Third-party', 'Play Now', 'Play the Demo']
df_games.loc[df_games['price'].isin(non_price_values), 'price'] = 0

Se calcula la media, la moda y la mediana de la columna price para imputar los nulos restantes.

In [76]:
media = df_games['price'].mean()
moda = df_games['price'].mode()[0]
mediana = df_games['price'].median()

print(f"La media de los precios de juegos Steam es {media:.2f}")
print(f"La moda de los precios de juegos Steam es {moda:.2f}")
print(f"La mediana de los precios de juegos Steam es {mediana:.2f}")

La media de los precios de juegos Steam es 8.84
La moda de los precios de juegos Steam es 4.99
La mediana de los precios de juegos Steam es 4.99


Inputamos los nulos con la mediana.

In [77]:
df_games['price'].fillna(df_games['price'].median(), inplace=True)
df_games.isnull().sum()

/tmp/ipykernel_250/4128492194.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_games['price'].fillna(df_games['price'].median(), inplace=True)
/tmp/ipykernel_250/4128492194.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_games['price'].fillna(df_games['price'].median(), inplace=True)


publisher          0
genres          3282
item_name          0
title              0
url                0
release_date       0
tags             162
reviews_url        0
specs            669
price              0
early_access       0
item_id            0
developer          0
is_equal           0
release_year       0
dtype: int64

#### Columnas de Generos, Tags y Especificaciones
- Gestión de nulos

In [78]:
df_games[df_games['genres'].isnull()][['genres','tags', 'specs']].sample(10)

genres                                               tags  \
7568     NaN                 [Martial Arts, Movie, Documentary]   
17938    NaN                                            [Movie]   
9889     NaN                                        [Indie, VR]   
10164    NaN  [Animation & Modeling, Design & Illustration, ...   
23156    NaN                              [Movie, Drama, Crime]   
6962     NaN                        [Design & Illustration, VR]   
8290     NaN  [Early Access, Action, Simulation, Indie, Casu...   
18371    NaN             [Early Access, Simulation, Sports, VR]   
23689    NaN                             [Anime, Movie, Action]   
15442    NaN                               [Casual, Simulation]   

                                                   specs  
7568                                [Captions available]  
17938                               [Captions available]  
9889   [Single-player, HTC Vive, Tracked Motion Contr...  
10164                                                NaN  
23156                               [Captions available]  
6962   [HTC Vive, Tracked Motion Controllers, Seated,...  
8290   [Single-player, Multi-player, Online Multi-Pla...  
18371  [Single-player, Steam Achievements, Steam Lead...  
23689                               [Captions available]  
15442  [Single-player, HTC Vive, Oculus Rift, Tracked...

In [79]:
df_games.genres.isnull().sum()

3282

In [80]:
# Exploramos el contenido de genres, tags y specs.

columns = ['genres', 'tags','specs']
df_games[columns].sample(10)

genres  \
7505                               NaN   
14073                          [Indie]   
10754                            [RPG]   
11398          [Indie, Racing, Sports]   
24776                  [Casual, Indie]   
16326  [Action, Adventure, Indie, RPG]   
14278             [Action, Indie, RPG]   
18189                              NaN   
30736                         [Action]   
13146                              NaN   

                                                    tags  \
7505   [VR, Casual, Indie, Massively Multiplayer, Adv...   
14073  [Bullet Hell, Great Soundtrack, Difficult, Fem...   
10754                                              [RPG]   
11398                            [Racing, Sports, Indie]   
24776                                    [Indie, Casual]   
16326  [RPG, Indie, Adventure, Action, Rogue-like, Pi...   
14278  [RPG, Action, Indie, Card Game, Singleplayer, ...   
18189      [Early Access, Simulation, Indie, Casual, VR]   
30736                                           [Action]   
13146                                [Indie, Sports, VR]   

                                                   specs  
7505   [Single-player, Multi-player, Online Multi-Pla...  
14073                                    [Single-player]  
10754  [Single-player, Downloadable Content, Full con...  
11398  [Single-player, Steam Achievements, Full contr...  
24776  [Single-player, Steam Achievements, Steam Trad...  
16326                [Single-player, Steam Achievements]  
14278  [Single-player, Steam Achievements, Full contr...  
18189  [Single-player, HTC Vive, Tracked Motion Contr...  
30736  [Single-player, Multi-player, Co-op, Downloada...  
13146  [Single-player, Local Multi-Player, Steam Achi...

Podemos observar que tags contiene tambien a genres por lo que podemos usarla para imputar los nulos en genres.

In [81]:
# Creamos una lista de géneros únicos
unique_genres = df_games['genres'].explode().unique()

# Definimos una función para buscar géneros en tags
def find_genres(row):
    if isinstance(row['tags'], list):
        genres_in_tags = [tag for tag in row['tags'] if tag in unique_genres]
        if genres_in_tags:
            return genres_in_tags
    return row['genres']

# Aplicamos la función a cada fila y almacenamos en la columna temporal 'genres_2'
df_games['genres_2'] = df_games.apply(find_genres, axis=1)
df_games[df_games['genres'].isnull()][['genres', 'genres_2', 'tags']].head()

genres                                          genres_2  \
4     NaN                   [Action, Indie, Casual, Sports]   
11    NaN                                          [Casual]   
19    NaN                             [Early Access, Indie]   
20    NaN  [Early Access, Action, Adventure, Indie, Casual]   
22    NaN   [Early Access, Strategy, Action, Indie, Casual]   

                                                 tags  
4                     [Action, Indie, Casual, Sports]  
11                                           [Casual]  
19                          [Early Access, Indie, VR]  
20   [Early Access, Action, Adventure, Indie, Casual]  
22  [Early Access, Strategy, Action, Indie, Casual...

Eliminamos la columna 'genres' para posteriormente reemplazarla por la columna 'genres_2'. y tambien se eliminará la columna 'tags' y la 'specs' porque no aportan información.

In [82]:

columns = ['genres', 'tags', 'specs']
df_games.drop(columns, axis=1, inplace=True)
# Renombramos la columna genres_2 como genres
df_games = df_games.rename(columns={'genres_2': 'genres'})
df_games.columns

Index(['publisher', 'item_name', 'title', 'url', 'release_date', 'reviews_url',
       'price', 'early_access', 'item_id', 'developer', 'is_equal',
       'release_year', 'genres'],
      dtype='object')

Remplazamos los valores nulos en la columna 'genres' por un valor unknown.

In [83]:
df_games['genres'] = df_games['genres'].apply(lambda x: ['unknown'] if isinstance(x, float) else x)
df_games.isnull().sum()

publisher       0
item_name       0
title           0
url             0
release_date    0
reviews_url     0
price           0
early_access    0
item_id         0
developer       0
is_equal        0
release_year    0
genres          0
dtype: int64

In [84]:
df_games = df_games.explode('genres')
df_games.head()

publisher            item_name                title  \
0  Kotoshiro  Lost Summoner Kitty  Lost Summoner Kitty   
0  Kotoshiro  Lost Summoner Kitty  Lost Summoner Kitty   
0  Kotoshiro  Lost Summoner Kitty  Lost Summoner Kitty   
0  Kotoshiro  Lost Summoner Kitty  Lost Summoner Kitty   
0  Kotoshiro  Lost Summoner Kitty  Lost Summoner Kitty   

                                                 url release_date  \
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   

                                         reviews_url  price early_access  \
0  http://steamcommunity.com/app/761140/reviews/?...   4.99        False   
0  http://steamcommunity.com/app/761140/reviews/?...   4.99        False   
0  http://steamcommunity.com/app/761140/reviews/?...   4.99        False   
0  http://steamcommunity.com/app/761140/reviews/?...   4.99        False   
0  http://steamcommunity.com/app/761140/reviews/?...   4.99        False   

  item_id  developer  is_equal release_year      genres  
0  761140  Kotoshiro      True         2018    Strategy  
0  761140  Kotoshiro      True         2018      Action  
0  761140  Kotoshiro      True         2018       Indie  
0  761140  Kotoshiro      True         2018      Casual  
0  761140  Kotoshiro      True         2018  Simulation

#### Columnas de URL y Reviews URL


In [85]:
import pyarrow

# Elimina las columnas 'url' y 'reviews_url' del dataframe.
columns = ['url', 'reviews_url']
df_games.drop(columns, axis=1, inplace=True)

# Cambia el tipo de dato de la columna 'early_access' a booleano.
df_games['early_access'] = df_games['early_access'].astype(bool)
df_games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 82806 entries, 0 to 32132
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   publisher     82806 non-null  object 
 1   item_name     82806 non-null  object 
 2   title         82806 non-null  object 
 3   release_date  82806 non-null  object 
 4   price         82806 non-null  float64
 5   early_access  82806 non-null  bool   
 6   item_id       82806 non-null  object 
 7   developer     82806 non-null  object 
 8   is_equal      82806 non-null  bool   
 9   release_year  82806 non-null  object 
 10  genres        82806 non-null  object 
dtypes: bool(2), float64(1), object(8)
memory usage: 6.5+ MB


# Carga

##### Cargamos el dataset con el dataset limpio y formato comprimido.

In [86]:
# Definimos la ruta de destino y el nombre del archivo.
folder_path = '../dataset/'
file_name = 'steam_games.parquet'

# Verificamos si el folder_path existe
if not os.path.exists(folder_path):
    # Si no existe, lo creamos
    os.makedirs(folder_path)

# Exportamos el DataFrame a un archivo Parquet
df_games.to_parquet(os.path.join(folder_path, file_name), engine='pyarrow', compression='zstd')

print(f'El archivo {file_name} se guardó correctamente en {folder_path}')

El archivo steam_games.parquet se guardó correctamente en ../dataset/
